src/main/sd/rw.c

```diff
@@ -0,0 +1,244 @@
+#include "sd.h"
+#include "debug.h"
+#include "pin.h"
+#include "spi.h"
+#include "command.h"
+
+STATIC uint8_t single_block_read(uint32_t index, block512_t* block)
+{
+    uint8_t command_argument[] = {(uint8_t)(index >> 24), (uint8_t)(index >> 16), (uint8_t)(index >> 8), (uint8_t)index};
+    mext2_command command = set_command(COMMAND_READ_SINGLE_BLOCK, command_argument);    //set CMD17
+
+    //prepare block for read
+    memset(block->data, 0xff, 512);
+
+    uint8_t buffer = 0xff;
+    wait_8_clock_cycles(&buffer);
+
+    mext2_response* response = send_command(&command, MEXT2_R1);
+    if(response == NULL || response -> r1 != 0)
+    {
+        mext2_error("Can't read block");
+        return 1;
+    }
+
+    if(wait_for_response((uint8_t*)response) == false)
+        return 1;
+
+    if(response -> r1 != 0xfe)
+    {
+        mext2_error("Received wrong data token");
+        return 1;
+    }
+
+    spi_read_write(block->data, 512);
+
+    wait_8_clock_cycles(&buffer);
+    wait_8_clock_cycles(&buffer);
+    wait_8_clock_cycles(&buffer);
+
+    return 0;
+}
+
+STATIC uint8_t multiple_block_read(uint32_t index, block512_t* block, uint8_t blocks_number)
+{
+    uint8_t command_argument[] = {(uint8_t)(index >> 24), (uint8_t)(index >> 16), (uint8_t)(index >> 8), (uint8_t)index};
+    mext2_command command = set_command(COMMAND_READ_MULTIPLE_BLOCK, command_argument);    //set CMD18
+
+    //prepare block for read
+    memset(block->data, 0xff, 512 * blocks_number);
+
+    uint8_t buffer = 0xff;
+    wait_8_clock_cycles(&buffer);
+
+    mext2_response* response = send_command(&command, MEXT2_R1);
+    if(response == NULL || response -> r1 != 0)
+    {
+        mext2_error("Can't read block");
+        return 1;
+    }
+
+    for(uint8_t i = 0; i < blocks_number; i++)
+    {
+        if(wait_for_response((uint8_t*)response) == false)
+            return 1;
+
+        if(response -> r1 != 0xfe)
+        {
+            mext2_error("Received wrong data token");
+            return 1;
+        }
+
+        spi_read_write(block[i].data, 512);
+
+        wait_8_clock_cycles(&buffer);
+        wait_8_clock_cycles(&buffer);
+    }
+
+    if(wait_for_response((uint8_t*)response) == false)
+        return 1;
+
+    memset(command_argument, 0x00, COMMAND_ARGUMENT_SIZE);
+    command = set_command(COMMAND_STOP_READ_DATA, command_argument);    //set CMD12
+
+    response = send_command(&command, MEXT2_R1b);
+    if(response == NULL || response -> r1 != 0)
+    {
+        mext2_error("Can't read block");
+        return 1;
+    }
+
+    wait_8_clock_cycles(&buffer);
+
+    buffer = 0;
+    for(uint8_t i = 0; buffer == 0; i++)
+    {
+        wait_8_clock_cycles(&buffer);
+    }
+
+    wait_8_clock_cycles(&buffer);
+    return 0;
+}
+
+uint8_t read_blocks(uint8_t blocks_number, uint32_t index, block512_t* block)
```

Comment:
Shouldn't we take here some sort of SD card descriptor in order to distinguish different versions of sd cards?
Sth like:
```suggestion
uint8_t read_blocks(SD* sd_card, uint8_t blocks_number, uint32_t index, block512_t* block)
```
Or should the higher layers care about this?